src/Bounds.cpp

```diff
@@ -1039,11 +1039,33 @@ class Bounds : public IRVisitor {
                     if (op->is_intrinsic(Call::shift_left)) {
                         if (t.is_int() && t.bits() >= 32) {
                             // Overflow is UB
-                            if (a_interval.has_lower_bound() && b_interval.has_lower_bound() && can_prove(b_interval.min >= 0 && b_interval.min < t.bits())) {
+                            // The following two Exprs are to avoid warning
+                            // messages that abs of an unsigned type is a no-op.
+                            Expr abs_b_interval_min =
+                                b_interval.min.type().is_int() ? abs(b_interval.min) : b_interval.min;
+                            Expr abs_b_interval_max =
+                                b_interval.max.type().is_int() ? abs(b_interval.max) : b_interval.max;
+                            if (a_interval.has_lower_bound() &&
+                                b_interval.has_lower_bound() &&
+                                can_prove(b_interval.min >= 0 &&
+                                          b_interval.min < t.bits())) {
                                 interval.min = a_interval.min << b_interval.min;
+                            } else if (a_interval.has_lower_bound() &&
+                                       b_interval.has_lower_bound() &&
+                                       can_prove(b_interval.min < 0 &&
+                                                 abs_b_interval_min < t.bits())) {
```

Comment:
Should this possibly be b_interval.min > -t.bits()? Then the abs_b_interval_min (and abs_b_interval_max) would be unused and could be removed. I'm also thinking maybe the simplifier will be more likely to be able to prove that than an expression with abs in it.

Reply:
Done